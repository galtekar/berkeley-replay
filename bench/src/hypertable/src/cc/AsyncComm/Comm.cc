/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

//#define DISABLE_LOG_DEBUG

#include <cassert>
#include <iostream>
using namespace std;

extern "C" {
#if defined(__APPLE__)
#include <arpa/inet.h>
#endif
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <poll.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <arpa/inet.h>
}

#include "Common/Error.h"
#include "Common/InetAddr.h"
#include "Common/FileUtils.h"
#include "Common/System.h"

#include "ReactorFactory.h"
#include "ReactorRunner.cc"
#include "Comm.h"
#include "IOHandlerAccept.h"
#include "IOHandlerData.h"

using namespace Hypertable;

atomic_t Comm::ms_next_request_id = ATOMIC_INIT(1);


/**
 */
Comm::Comm() {
  if (ReactorFactory::ms_reactors.size() == 0) {
    HT_ERROR("ReactorFactory::initialize must be called before creating AsyncComm::comm object");
    DUMP_CORE;
  }
  ReactorFactory::get_reactor(m_timer_reactor_ptr);
  m_handler_map_ptr = ReactorRunner::ms_handler_map_ptr;
}



/**
 *
 */
Comm::~Comm() {

  set<IOHandler *> handlers;
  m_handler_map_ptr->decomission_all(handlers);
  for (set<IOHandler *>::iterator iter = handlers.begin(); iter != handlers.end(); ++iter)
    (*iter)->shutdown();

  // wait for all decomissioned handlers to get purged by Reactor
  // (I think this is no longer needed, since the reactors should have been
  //  shut down before destroying the Comm layer)
  // m_handler_map_ptr->wait_for_empty();
}



/**
 */
int Comm::connect(struct sockaddr_in &addr, DispatchHandlerPtr &default_handler_ptr) {
  int sd;

  if (m_handler_map_ptr->contains_handler(addr))
    return Error::COMM_ALREADY_CONNECTED;

  if ((sd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
    cerr << "socket() failure: " << strerror(errno) << endl;
    exit(1);
  }

  return connect_socket(sd, addr, default_handler_ptr);
}



/**
 *
 */
int Comm::connect(struct sockaddr_in &addr, struct sockaddr_in &local_addr, DispatchHandlerPtr &default_handler_ptr) {
  int sd;

  if (m_handler_map_ptr->contains_handler(addr))
    return Error::COMM_ALREADY_CONNECTED;

  if ((sd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
    HT_ERRORF("socket() failure: %s", strerror(errno));
    exit(1);
  }

  // bind socket to local address
  if ((bind(sd, (const sockaddr *)&local_addr, sizeof(sockaddr_in))) < 0) {
    HT_ERRORF("bind() failure: %s", strerror(errno));
    exit(1);
  }

  return connect_socket(sd, addr, default_handler_ptr);
}


/**
 *
 */
int Comm::listen(struct sockaddr_in &addr, ConnectionHandlerFactoryPtr &chf_ptr) {
  DispatchHandlerPtr nullHandlerPtr(0);
  return listen(addr, chf_ptr, nullHandlerPtr);
}


/**
 * 
 */
int Comm::set_alias(struct sockaddr_in &addr, struct sockaddr_in &alias) {
  boost::mutex::scoped_lock lock(m_mutex);
  return m_handler_map_ptr->set_alias(addr, alias);
}


/**
 *
 */
int Comm::listen(struct sockaddr_in &addr, ConnectionHandlerFactoryPtr &chf_ptr, DispatchHandlerPtr &default_handler_ptr) {
  IOHandlerPtr handlerPtr;
  IOHandlerAccept *acceptHandler;
  int one = 1;
  int sd;

  if ((sd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
    cerr << "socket() failure: " << strerror(errno) << endl;
    exit(1);
  }

#if defined(__linux__)
  if (setsockopt(sd, SOL_TCP, TCP_NODELAY, &one, sizeof(one)) < 0)
    cerr << "setsockopt(TCP_NODELAY) failure: " << strerror(errno) << endl;
#elif defined(__APPLE__)
  if (setsockopt(sd, SOL_SOCKET, SO_NOSIGPIPE, &one, sizeof(one)) < 0)
    HT_WARNF("setsockopt(SO_NOSIGPIPE) failure: %s", strerror(errno));
#endif

  if (setsockopt(sd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one)) < 0)
    cerr << "setsockopt(SO_REUSEADDR) failure: " << strerror(errno) << endl;

  // bind socket
  if ((bind(sd, (const sockaddr *)&addr, sizeof(sockaddr_in))) < 0) {
    cerr << "bind() failure: " << strerror(errno) << endl;
    exit(1);
  }

  if (::listen(sd, 64) < 0) {
    cerr << "listen() failure: " << strerror(errno) << endl;
    exit(1);
  }

  handlerPtr = acceptHandler = new IOHandlerAccept(sd, addr, default_handler_ptr, m_handler_map_ptr, chf_ptr);
  m_handler_map_ptr->insert_handler(acceptHandler);
  acceptHandler->start_polling();

  return Error::OK;
}



int Comm::send_request(struct sockaddr_in &addr, time_t timeout, CommBufPtr &cbuf_ptr, DispatchHandler *responseHandler) {
  boost::mutex::scoped_lock lock(m_mutex);
  IOHandlerDataPtr dataHandlerPtr;
  Header::HeaderT *mheader = (Header::HeaderT *)cbuf_ptr->data;
  int error = Error::OK;

  cbuf_ptr->reset_data_pointers();

  if (!m_handler_map_ptr->lookup_data_handler(addr, dataHandlerPtr)) {
    HT_ERRORF("No connection for %s:%d", inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));
    return Error::COMM_NOT_CONNECTED;
  }

  mheader->flags |= Header::FLAGS_BIT_REQUEST;
  if (responseHandler == 0)
    mheader->flags |= Header::FLAGS_BIT_IGNORE_RESPONSE;
  mheader->id = atomic_inc_return(&ms_next_request_id);
  if (mheader->id == 0)
    mheader->id = atomic_inc_return(&ms_next_request_id);

  if ((error = dataHandlerPtr->send_message(cbuf_ptr, timeout, responseHandler)) != Error::OK)
    dataHandlerPtr->shutdown();

  return error;
}


int Comm::send_response(struct sockaddr_in &addr, CommBufPtr &cbuf_ptr) {
  boost::mutex::scoped_lock lock(m_mutex);
  IOHandlerDataPtr dataHandlerPtr;
  Header::HeaderT *mheader = (Header::HeaderT *)cbuf_ptr->data;
  int error = Error::OK;

  cbuf_ptr->reset_data_pointers();

  if (!m_handler_map_ptr->lookup_data_handler(addr, dataHandlerPtr)) {
    HT_ERRORF("No connection for %s:%d", inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));
    return Error::COMM_NOT_CONNECTED;
  }

  mheader->flags &= Header::FLAGS_MASK_REQUEST;

  if ((error = dataHandlerPtr->send_message(cbuf_ptr)) != Error::OK)
    dataHandlerPtr->shutdown();

  return error;
}



/**
 * 
 */
int Comm::create_datagram_receive_socket(struct sockaddr_in *addr, DispatchHandlerPtr &dispatchHandlerPtr) {
  IOHandlerPtr handlerPtr;
  IOHandlerDatagram *datagramHandler;
  int one = 1;
  int sd;

  if ((sd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
    HT_ERRORF("socket() failure: %s", strerror(errno));
    exit(1);
  }

  // Set to non-blocking
  FileUtils::set_flags(sd, O_NONBLOCK);

  int bufsize = 4*32768;
  if (setsockopt(sd, SOL_SOCKET, SO_SNDBUF, (char *)&bufsize, sizeof(bufsize)) < 0) {
    HT_ERRORF("setsockopt(SO_SNDBUF) failed - %s", strerror(errno));
  }
  if (setsockopt(sd, SOL_SOCKET, SO_RCVBUF, (char *)&bufsize, sizeof(bufsize)) < 0) {
    HT_ERRORF("setsockopt(SO_RCVBUF) failed - %s", strerror(errno));
  }

  if (setsockopt(sd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one)) < 0) {
    HT_WARNF("setsockopt(SO_REUSEADDR) failure: %s", strerror(errno));
  }

  // bind socket
  if ((bind(sd, (const sockaddr *)addr, sizeof(sockaddr_in))) < 0) {
    HT_ERRORF("bind() failure: %s", strerror(errno));
    exit(1);
  }

  handlerPtr = datagramHandler = new IOHandlerDatagram(sd, *addr, dispatchHandlerPtr);

  handlerPtr->get_local_address(addr);

  m_handler_map_ptr->insert_datagram_handler(datagramHandler);
  datagramHandler->start_polling();

  return Error::OK;
}


/**
 * 
 */
int Comm::send_datagram(struct sockaddr_in &addr, struct sockaddr_in &send_addr, CommBufPtr &cbuf_ptr) {
  boost::mutex::scoped_lock lock(m_mutex);
  IOHandlerDatagramPtr datagramHandlerPtr;
  Header::HeaderT *mheader = (Header::HeaderT *)cbuf_ptr->data;
  int error = Error::OK;

  cbuf_ptr->reset_data_pointers();

  if (!m_handler_map_ptr->lookup_datagram_handler(send_addr, datagramHandlerPtr)) {
    std::string str;
    HT_ERRORF("Datagram send address %s not registered", InetAddr::string_format(str, send_addr));
    DUMP_CORE;
  }

  mheader->flags |= (Header::FLAGS_BIT_REQUEST|Header::FLAGS_BIT_IGNORE_RESPONSE);

  if ((error = datagramHandlerPtr->send_message(addr, cbuf_ptr)) != Error::OK)
    datagramHandlerPtr->shutdown();

  return error;
}



/**
 *
 */
int Comm::set_timer(uint64_t duration_millis, DispatchHandler *handler) {
  struct TimerT timer;
  boost::xtime_get(&timer.expireTime, boost::TIME_UTC);
  timer.expireTime.sec += duration_millis / 1000LL;
  timer.expireTime.nsec += (duration_millis % 1000LL) * 1000000LL;
  timer.handler = handler;
  m_timer_reactor_ptr->add_timer(timer);
  return Error::OK;
}



/**
 *
 */
int Comm::set_timer_absolute(boost::xtime expire_time, DispatchHandler *handler) {
  struct TimerT timer;  
  memcpy(&timer.expireTime, &expire_time, sizeof(boost::xtime));
  timer.handler = handler;
  m_timer_reactor_ptr->add_timer(timer);
  return Error::OK;
}

/**
 *
 */
int Comm::get_local_address(struct sockaddr_in addr, struct sockaddr_in *local_addr) {
  boost::mutex::scoped_lock lock(m_mutex);
  IOHandlerDataPtr dataHandlerPtr;

  if (!m_handler_map_ptr->lookup_data_handler(addr, dataHandlerPtr)) {
    HT_ERRORF("No connection for %s:%d", inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));
    return Error::COMM_NOT_CONNECTED;
  }

  dataHandlerPtr->get_local_address(local_addr);

  return Error::OK;
}


/**
 * 
 */
int Comm::close_socket(struct sockaddr_in &addr) {
  IOHandlerPtr handlerPtr;

  if (!m_handler_map_ptr->decomission_handler(addr, handlerPtr))
    return Error::COMM_NOT_CONNECTED;

  handlerPtr->shutdown();

  return Error::OK;
}


/**
 *  ----- Private methods -----
 */


/**
 *
 */
int Comm::connect_socket(int sd, struct sockaddr_in &addr, DispatchHandlerPtr &default_handler_ptr) {
  IOHandlerPtr handlerPtr;
  IOHandlerData *dataHandler;
  int one = 1;

  // Set to non-blocking
  FileUtils::set_flags(sd, O_NONBLOCK);

#if defined(__linux__)
  if (setsockopt(sd, SOL_TCP, TCP_NODELAY, &one, sizeof(one)) < 0)
    cerr << "setsockopt(TCP_NODELAY) failure: " << strerror(errno) << endl;
#elif defined(__APPLE__)
  if (setsockopt(sd, SOL_SOCKET, SO_NOSIGPIPE, &one, sizeof(one)) < 0)
    HT_WARNF("setsockopt(SO_NOSIGPIPE) failure: %s", strerror(errno));
#endif

  handlerPtr = dataHandler = new IOHandlerData(sd, addr, default_handler_ptr);
  m_handler_map_ptr->insert_handler(dataHandler);

  while (::connect(sd, (struct sockaddr *)&addr, sizeof(struct sockaddr_in)) < 0) {
    if (errno == EINTR) {
      poll(0, 0, 1000);
      continue;
    }
    else if (errno == EINPROGRESS) {
      dataHandler->start_polling();
      dataHandler->add_poll_interest(Reactor::READ_READY|Reactor::WRITE_READY);
      return Error::OK;
    }
    HT_ERRORF("connect() failure : %s", strerror(errno));
    exit(1);
  }

  dataHandler->start_polling();

  return Error::OK;
}
