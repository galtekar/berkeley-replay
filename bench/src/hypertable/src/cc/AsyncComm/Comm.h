/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef HYPERTABLE_COMMENGINE_H
#define HYPERTABLE_COMMENGINE_H

#include <string>

#include <boost/thread/mutex.hpp>
#include <boost/thread/xtime.hpp>

extern "C" {
#include <stdint.h>
}

#include "Common/ReferenceCount.h"

#include "DispatchHandler.h"
#include "CommBuf.h"
#include "ConnectionHandlerFactory.h"
#include "HandlerMap.h"

using namespace std;

namespace Hypertable {

  /**
   * Provides communication (message passing) services to an application.
   * There should be only one instance of this class per
   * process and the static method ReactorFactory#initialize must be called
   * prior to constructing this class in order to create the system-wide
   * I/O reactor threads.
   */
  class Comm : public ReferenceCount {

  public:

    Comm();

    ~Comm();

    /**
     * Establishes a TCP connection to the address given by the addr argument
     * and associates with it a default dispatch handler.  CONNECTION_ESTABLISHED
     * and DISCONNECT events are delivered to the default dispatch handler.  The
     * argument addr is used to subsequently refer to the connection.
     *
     * @param addr IP address and port to connect to
     * @param default_handler_ptr smart pointer to default dispatch handler
     * @return Error::OK on success or error code on failure
     */
    int connect(struct sockaddr_in &addr, DispatchHandlerPtr &default_handler_ptr);

    /**
     * Establishes a TCP connection to the address given by the addr argument,
     * binding the local side of the connection to the address given by the
     * local_addr argument.  A default dispatch handler is associated with the
     * connection to receive CONNECTION_ESTABLISHED and DISCONNECT events.  The
     * argument addr is used to subsequently refer to the connection.
     *
     * @param addr IP address and port to connect to
     * @param local_addr Local IP address and port to bind to
     * @param default_handler_ptr smart pointer to default dispatch handler
     * @return Error::OK on success or error code on failure
     */
    int connect(struct sockaddr_in &addr, struct sockaddr_in &local_addr, DispatchHandlerPtr &default_handler_ptr);

    /**
     * Sets an alias for a TCP connection
     *
     * @param addr connection identifier (remote address)
     * @param alias alias connection identifier
     * @return Error::OK on success or error code on failure
     */
    int set_alias(struct sockaddr_in &addr, struct sockaddr_in &alias);

    /**
     * Tells the communication subsystem to listen for connection requests on the
     * address given by the addr argument.  New connections will be assigned
     * dispatch handlers by invoking the get_instance method of the connection
     * handler factory supplied as the chf_ptr argument.  CONNECTION_ESTABLISHED
     * events are logged, but not delivered to the application
     *
     * @param addr IP address and port to listen for connection on
     * @param chf_ptr connection handler factory smart pointer
     * @return Error::OK on success or error code on failure
     */
    int listen(struct sockaddr_in &addr, ConnectionHandlerFactoryPtr &chf_ptr);

    /**
     * Tells the communication subsystem to listen for connection requests on the
     * address given by the addr argument.  New connections will be assigned
     * dispatch handlers by invoking the get_instance method of the connection
     * handler factory supplied as the chf_ptr argument.  CONNECTION_ESTABLISHED
     * events are delivered via the default dispatch handler supplied in the
     * default_handler_ptr argument.
     *
     * @param addr IP address and port to listen for connection on
     * @param chf_ptr connection handler factory smart pointer
     * @param default_handler_ptr smart pointer to default dispatch handler
     * @return Error::OK on success or error code on failure
     */
    int listen(struct sockaddr_in &addr, ConnectionHandlerFactoryPtr &chf_ptr, DispatchHandlerPtr &default_handler_ptr);

    /**
     * Sends a request message over a connection, expecting a response.
     * The connection is specified by the addr argument which is the remote
     * end of the connection.  The request message to send is encapsulated
     * in the cbuf_ptr object (see CommBuf) and should start with a valid header.
     * The dispatch handler given by the response_handler argument will get
     * called back with either a response MESSAGE event or a TIMEOUT event if     
     * no response is received within the number of seconds specified by the
     * timeout argument.
     *
     * <p>If the server at the other end of the connection uses an ApplicationQueue
     * to carry out requests, then the gid field in the header can be used to
     * serialize requests that are destined for the same object.  For example,
     * the following code serializes requests to the same file descriptor:
     *
     * <pre>
     * HeaderBuilder hbuilder(Header::PROTOCOL_DFSBROKER);
     * hbuilder.set_group_id(fd);
     * CommBuf *cbuf = new CommBuf(hbuilder, 14);
     * cbuf->AppendShort(COMMAND_READ);
     * cbuf->AppendInt(fd);
     * cbuf->AppendLong(amount);
     * </pre>
     *
     * @param addr connection identifier (remote address)
     * @param timeout number of seconds to wait before delivering TIMEOUT event
     * @param cbuf_ptr request message to send (see CommBuf)
     * @param response_handler pointer to response handler associated with the request
     * @return Error::OK on success or error code on failure
     */
    int send_request(struct sockaddr_in &addr, time_t timeout, CommBufPtr &cbuf_ptr, DispatchHandler *response_handler);

    /**
     * Sends a response message back over a connection.  It is assumed that the
     * id field of the header matches the id field of the request for which this
     * is a response to.  The connection is specified by the addr argument which
     * is the remote end of the connection.  The response message to send is
     * encapsulated in the cbuf_ptr (see CommBuf) object and should start with a
     * valid header.  The following code snippet illustrates how a simple response
     * message gets created to send back to a client in response to a request
     * message:
     *
     * <pre>
     * HeaderBuilder hbuilder;
     * hbuilder.initialize_from_request(request_event_ptr->header);
     * CommBufPtr cbufPtr( new CommBuf(hbuilder, 4) );
     * cbufPtr->append_int(Error::OK);
     * </pre>
     *
     * @param addr connection identifier (remote address)
     * @param cbuf_ptr response message to send (must have valid header with matching request id)
     * @return Error::OK on success or error code on failure
     */
    int send_response(struct sockaddr_in &addr, CommBufPtr &cbuf_ptr);

    /**
     * Obtains the local address of a socket connection.  The connection is
     * identified by the remote address in the addr argument.
     *
     * @param addr connection identifier (remote address)
     * @param local_addr pointer to address structure to hold the returned local address
     * @return Error::OK on success or error code on failure
     */
    int get_local_address(struct sockaddr_in addr, struct sockaddr_in *local_addr);

    /**
     * Creates a local socket for receiving datagrams and assigns a default dispatch
     * handler to handle events on this socket.  This socket can also be used for
     * sending datagrams.  The events delivered for this socket consist of either MESSAGE
     * events or ERROR events.
     *
     * @param addr address of the socket
     * @param handler_ptr default dispatch handler to handle the deliver of events
     * @return Error::OK on success or error code on failure
     */
    int create_datagram_receive_socket(struct sockaddr_in *addr, DispatchHandlerPtr &handler_ptr);

    /**
     * Sends a datagram to a remote address.  The remote address is specified by the addr
     * argument and the local socket address to send it from is specified by the send_addr
     * argument.  The send_addr argument must refer to a socket that was created with a call
     * to <a href="#create_datagram_receive_socket">create_datagram_receive_socket</a>.
     *
     * @param addr remote address to send datagram to
     * @param send_addr local socket address to send from
     * @param cbuf_ptr datagram message with valid header
     * @return Error::OK on success or error code on failure
     */
    int send_datagram(struct sockaddr_in &addr, struct sockaddr_in &send_addr, CommBufPtr &cbuf_ptr);

    /**
     * Sets a timer that will generate a TIMER event after some number of milliseconds
     * have elapsed, specified by the duration_millis argument.  The handler argument
     * represents the dispatch handler to receive the TIMER event.
     *
     * @param duration_millis number of milliseconds to wait
     * @param handler the dispatch handler to receive the TIMER event upon expiration
     * @return Error::OK on success or error code on failure
     */
    int set_timer(uint64_t duration_millis, DispatchHandler *handler);

    /**
     * Sets a timer that will generate a TIMER event at the absolute time specified
     * by the expire_time argument.  The handler argument represents the dispatch
     * handler to receive the TIMER event.
     *
     * @param expire_time number of milliseconds to wait
     * @param handler the dispatch handler to receive the TIMER event upon expiration
     * @return Error::OK on success or error code on failure
     */
    int set_timer_absolute(boost::xtime expire_time, DispatchHandler *handler);

    /**
     * Closes the socket connection specified by the addr argument.  This has the 
     * effect of closing the connection and removing it from the event demultiplexer
     * (e.g epoll).  It also causes all outstanding requests on the connection to
     * get purged.
     *
     * @param addr address of socket connection to close
     * @return Error::OK on success or error code on failure
     */
    int close_socket(struct sockaddr_in &addr);

  private:

    int connect_socket(int sd, struct sockaddr_in &addr, DispatchHandlerPtr &default_handler_ptr);

    static atomic_t ms_next_request_id;

    boost::mutex   m_mutex;
    HandlerMapPtr  m_handler_map_ptr;
    ReactorPtr     m_timer_reactor_ptr;
  };
  typedef boost::intrusive_ptr<Comm> CommPtr;

}

#endif // HYPERTABLE_COMMENGINE_H
