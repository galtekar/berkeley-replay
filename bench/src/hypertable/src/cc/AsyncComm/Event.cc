/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */


#include <iostream>
using namespace std;

extern "C" {
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
}

#include "Common/Error.h"
#include "Common/StringExt.h"

#include "Event.h"

namespace Hypertable {

  std::string Event::toString() {
    string dstr;

    dstr = "Event: type=";
    if (type == CONNECTION_ESTABLISHED)
      dstr += "CONNECTION_ESTABLISHED";
    else if (type == DISCONNECT)
      dstr += "DISCONNECT";
    else if (type == MESSAGE) {
      dstr += "MESSAGE protocol=";
      if (header->protocol < Header::PROTOCOL_MAX)
	dstr += Header::protocolStrings[header->protocol];
      else
	dstr += "unknown";
      dstr += (std::string)" id=" + (int)header->id;
      dstr += (std::string)" gid=" + (int)header->gid;
      dstr += (std::string)" headerLen=" + (int)header->headerLen;
      dstr += (std::string)" totalLen=" + (int)header->totalLen;
    }
    else if (type == TIMER)
      dstr += "TIMER";
    else if (type == ERROR)
      dstr += "ERROR";
    else
      dstr += (int)type;

    if (error != Error::OK)
      dstr += (std::string)" \"" + Error::get_text(error) + "\"";

    if (type != TIMER) {
      dstr += " from=";
      dstr += (std::string)inet_ntoa(addr.sin_addr) + ":" + (int)ntohs(addr.sin_port);
    }

    return dstr;
  }

}


