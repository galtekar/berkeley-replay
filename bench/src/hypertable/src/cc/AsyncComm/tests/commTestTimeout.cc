/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <cstdlib>

extern "C" {
#include <poll.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>
}

#include <boost/thread/thread.hpp>

#include "Common/Error.h"
#include "Common/FileUtils.h"
#include "Common/InetAddr.h"
#include "Common/Logger.h"
#include "Common/TestHarness.h"
#include "Common/System.h"
#include "Common/Usage.h"

#include "AsyncComm/Comm.h"
#include "AsyncComm/ConnectionManager.h"
#include "AsyncComm/Event.h"
#include "AsyncComm/ReactorFactory.h"

#include "CommTestThreadFunction.h"

using namespace Hypertable;

namespace {
  const char *usage[] = {
    "usage: commTestTimout",
    "",
    "This program tests tests the request timeout logic of AsyncComm.",
    0
  };

  const int DEFAULT_PORT = 32998;
  const char *DEFAULT_PORT_ARG = "--port=32998";

  class ServerLauncher {
  public:
    ServerLauncher() {
      if ((m_child_pid = fork()) == 0) {
	execl("./testServer", "./testServer", DEFAULT_PORT_ARG, "--delay=120000", (char *)0);
      }
      poll(0,0,2000);
    }
    ~ServerLauncher() {
      if (kill(m_child_pid, 9) == -1)
	perror("kill");
    }
    private:
      pid_t m_child_pid;
  };

  /**
   * 
   */
  class ResponseHandler : public DispatchHandler {
  public:
    ResponseHandler() : m_mutex(), m_cond(), m_connected(false) { return; }

    virtual void handle(EventPtr &eventPtr) {
      boost::mutex::scoped_lock lock(m_mutex);
      if (eventPtr->type == Event::CONNECTION_ESTABLISHED) {
	m_connected = true;
	m_cond.notify_one();
      }
      else if (eventPtr->type == Event::ERROR) {
	HT_INFOF("%s", eventPtr->toString().c_str());
      }
      else {
	HT_INFOF("%s", eventPtr->toString().c_str());
	m_connected = true;
	m_cond.notify_one();
      }
    }

    void wait_for_connection() {
      boost::mutex::scoped_lock lock(m_mutex);
      while (!m_connected)
	m_cond.wait(lock);
    }

  private:
    boost::mutex      m_mutex;
    boost::condition  m_cond;
    bool              m_connected;
  };

}


int main(int argc, char **argv) {
  struct sockaddr_in addr;
  Comm *comm;
  int error;
  EventPtr eventPtr;
  TestHarness harness("commTestTimeout");
  bool golden = false;
  ResponseHandler *respHandler = new ResponseHandler();
  DispatchHandlerPtr dispatchHandlerPtr(respHandler);

  {
    ServerLauncher slauncher;

    if (argc > 1) {
      if (!strcmp(argv[1], "--golden"))
	golden = true;
      else
	Usage::dump_and_exit(usage);
    }

    srand(8876);

    System::initialize(argv[0]);
    ReactorFactory::initialize(1);

    InetAddr::initialize(&addr, "localhost", DEFAULT_PORT);

    comm = new Comm();

    if ((error = comm->connect(addr, dispatchHandlerPtr)) != Error::OK)
      return 1;

    respHandler->wait_for_connection();

    HeaderBuilder hbuilder(Header::PROTOCOL_NONE, rand());
    std::string msg;

    msg = "foo";
    CommBufPtr cbufPtr( new CommBuf(hbuilder, Serialization::encoded_length_string(msg)) );
    cbufPtr->append_string(msg);
    if ((error = comm->send_request(addr, 5, cbufPtr, respHandler)) != Error::OK) {
      HT_ERRORF("Problem sending request - %s", Error::get_text(error));
      return 1;
    }

    msg = "bar";
    cbufPtr.reset (new CommBuf(hbuilder, Serialization::encoded_length_string(msg)) );
    cbufPtr->append_string(msg);
    if ((error = comm->send_request(addr, 5, cbufPtr, respHandler)) != Error::OK) {
      HT_ERRORF("Problem sending request - %s", Error::get_text(error));
      return 1;
    }

    poll(0, 0, 8000);

    delete comm;

  }

  if (!golden)
    harness.validate_and_exit("commTestTimeout.golden");

  harness.regenerate_golden_file("commTestTimeout.golden");

  return 0;
}
