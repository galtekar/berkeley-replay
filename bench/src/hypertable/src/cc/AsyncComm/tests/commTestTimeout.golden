INFO commTestTimeout : (${SRC_DIR}/commTestTimeout.cc:90) Event: type=ERROR "COMM request timeout" from=127.0.0.1:32998
INFO commTestTimeout : (${SRC_DIR}/commTestTimeout.cc:90) Event: type=ERROR "COMM request timeout" from=127.0.0.1:32998
